### PR TITLE
Added i18n feature for dynamic state descriptions

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added i18n feature for dynamic state descriptions via a `BaseDynamicStateDescriptionProvider` class
- Added null annotations to `StateOption`, `StateDescription` and `ChannelStateDescriptionProvider`
- Make use of `StateDescriptionFragmentBuilder`
- Changed null annotation of `CommandOption` label as we always allow `Nullable` labels
- Minor improvements

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

**Comment by cweitkamp** [comment]
What is wrong with the itest here?

```
[ERROR] Error   : Exit code remote process 1: java -cp /home/travis/build/openhab/openhab-core/itests/org.openhab.core.thing.tests/target/test/tmp/testing/itest/cnf/cache/4.2.0/org.openhab.core.bom.runtime-index/org.eclipse.osgi-3.13.100.v20180827-1536.jar:/home/travis/build/openhab/openhab-core/itests/org.openhab.core.thing.tests/target/test/tmp/testing/itest/cnf/cache/4.2.0/bnd-cache/biz.aQute.launcher/biz.aQute.launcher-4.2.0.jar -Dlauncher.properties="/home/travis/build/openhab/openhab-core/itests/org.openhab.core.thing.tests/target/test/tmp/testing/itest/generated/launch3088335997198286107.properties" -ea aQute.launcher.Launcher
```

Is it my fault? I cannot find the reason for it.

**Comment by cweitkamp** [comment]
@maggu2810 Is there a specific reason why the tests for the Magic bundle are commented out in the `itests/pom.xml`?

https://github.com/openhab/openhab-core/blob/71b241ae21a89ae67eb2b935aaf64497c632c928/itests/pom.xml#L48

**Comment by maggu2810** [comment]
No technical one, just not enough time...

IIRC the comment out modules has not yet been migrated -- or at least tested.
IIRC I just applied some common migration steps for all bundles.
After that I checked what is working out of the box and enabled them.
After that I invest some time and fixed some other ones so they are working.

So, feel free to have a look at the magic test (I assume also pure unit tests needs to be moved manually).

**Comment by cweitkamp** [comment]
The unit test for Magic Bundle are working now. But from my point of view there are no integration tests yet. I guess we could move all existing tests.

File: bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        bundleContext = componentContext.getBundleContext();
+    }
+
+    @Deactivate
+    public void deactivate() {
+        channelOptionsMap.clear();
+        bundleContext = null;
+    }
+
+    @Reference
```
Comment (maggu2810) [inline bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: Do not depend on DS annotation inheritance (see e.g. https://community.liferay.com/blogs/-/blogs/bnd-instruction-to-avoid -- it has already been discussion on ESH some time).
If a component use this class a parent, he should define the annotated methods itself and call the parent class implementation.

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: It somehow makes sense. A subclass has to add a `@Component` annotation too to provide its service. I will update the class accordingly but will remove the setter and getter completely. Similar to the `AbstractDiscoveryService` - I always was wondering why it holds instances of two services without requiring them. That is clear now.

File: itests/org.openhab.core.thing.tests/src/main/java/org/eclipse/smarthome/core/thing/type/ChannelStateDescriptionProviderOSGiTest.java [hunk 2]
```diff
@@ -77,14 +81,20 @@
...
     private ItemChannelLinkRegistry linkRegistry;
 
     @Mock
     private ComponentContext componentContext;
 
+    private Bundle testBundle;
+
     @Before
-    public void setup() {
+    public void setup() throws Exception {
```
Comment (davidgraeff) [inline itests/org.openhab.core.thing.tests/src/main/java/org/eclipse/smarthome/core/thing/type/ChannelStateDescriptionProviderOSGiTest.java]: That's very generic ;)

Comment (cweitkamp) [reply itests/org.openhab.core.thing.tests/src/main/java/org/eclipse/smarthome/core/thing/type/ChannelStateDescriptionProviderOSGiTest.java]: Unfortunately, I have to

https://github.com/openhab/openhab-core/blob/00edae4539f9ea4188d34afb68424ac8cfb33232/bundles/org.openhab.core.test/src/main/java/org/eclipse/smarthome/test/SyntheticBundleInstaller.java#L79-L92

**Comment by maggu2810** [comment]
My mail UI didn't tell me that the WIP has been removed long time ago.
I will add this PR to my todo list.

**Comment by maggu2810** [comment]
There are merge conflicts.
Would it make sense to split the "enable itests of magic bundle" changes to a separate PR?

**Comment by cweitkamp** [comment]
> Would it make sense to split the "enable itests of magic bundle" changes to a separate PR?

Yes, I thought about that too. Give me a minute.

**Comment by cweitkamp** [comment]
Done: Split-up PR into #756 and this one.

**Comment by maggu2810** [comment]
There is still a merge conflict.

**Comment by cweitkamp** [comment]
Rebased again.

File: bundles/org.openhab.core.test.magic/src/main/resources/ESH-INF/thing/channel-types.xml [hunk 3]
```diff
@@ -91,6 +92,11 @@
 		<autoUpdatePolicy>recommend</autoUpdatePolicy>
 	</channel-type>
 
+	<channel-type id="systemcommand">
+		<label>Send system command</label>
+		<description>Sends a system command to Kodi. This allows to shutdown/suspend/hibernate/reboot/quit Kodi.</description>
```
Comment (maggu2810) [inline bundles/org.openhab.core.test.magic/src/main/resources/ESH-INF/thing/channel-types.xml]: I really don't care, but I don't think there is some interaction with code. Is there? :wink:

**Comment by maggu2810** [comment]
Magic bundle test failed on Travis CI.

File: bundles/org.openhab.core.test.magic/src/main/resources/ESH-INF/thing/channel-types.xml (hunk 3 above)
Comment (cweitkamp) [reply bundles/org.openhab.core.test.magic/src/main/resources/ESH-INF/thing/channel-types.xml]: No, just copy & paste from the binding where the issue has its origin.

**Comment by cweitkamp** [comment]
Alright. Green again.

File: bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(Channel channel, @Nullable StateDescription original,
+            @Nullable Locale locale) {
+        // can be overridden by subclasses
+        ChannelUID channelUID = channel.getUID();
+        String pattern = channelPatternMap.get(channelUID);
+        List<StateOption> options = channelOptionsMap.get(channelUID);
+        if (pattern == null && options == null) {
```
Comment (Hilbrand) [inline bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: This suggests both `setStatePattern` and `setStateOptions` need to be set. Given the javadoc of `setStatePattern` that if not set the default pattern is used I would suspect if `pattern` is null here first `original.getPattern()` should be used and checked if that isn't null either?

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: I am not sure if I get your point. The `StateDescriptionProvider` only should return an new `StateDescription` instance if at least one values is set. In general the `StateDescriptionService` is responsible for merging it with the original `StateDescription`, so we do not need not take care of that here.

Comment (Hilbrand) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: Ok. I don't know what I was thinking here, maybe I shouldn't write code when tired :smile:. This code seems fine. So I'll have to search further why my other changes didn't work...

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: Never mind. I know that :laugh:.

Do you already created a PR (or have a piece of code) to point me to? I can have a look too. And I have some idea related to the component declaration of a `StateDescriptionProvider` which I like to discuss with you. It does not really fit in here.

Comment (Hilbrand) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: I haven't created a pr, but the code is here: https://github.com/Hilbrand/openhab2-addons/commits/spotif_base_dynamic
I tested it locally, but not a lot. It doesn't show the dynamic lists anymore, so I started debugging, but not a lot. Interested in your idea.

Comment (cweitkamp) [reply bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseDynamicStateDescriptionProvider.java]: Thanks for the link. See my https://github.com/Hilbrand/openhab2-addons/commit/9c4f586cf6bb64e3cd0c229290a5641e38113521 over there.

**Pull request merged**